File: leeaa/17.py
class Solution:
    def letterCombinations(self, digits):
        arr = ["","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"]
        if len(digits) == 0:
            return []
        

        stack = [0]*len(digits)
        ans = []
        while len(stack):
            #print(ans)
            #print(stack)
            if len(stack)==len(digits):
                tmpc = []
                for i in range(len(stack)-1):
                    tmpc.append(arr[int(digits[i])][stack[i]])
                for i in arr[int(digits[-1])]:
                    ans.append(''.join(tmpc + [i]))
                stack.pop()
            else:
                if stack[-1] + 1 < len(arr[int(digits[len(stack)-1])]):
                    stack[-1] += 1
                    while len(stack) < len(digits):
                        stack.append(0)
                else:
                    stack.pop()
        return ans

A = Solution()
print(A.letterCombinations("23"))

File: rein/17.js
/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function(digits) {
    if (digits.length === 0) {
        return [];
    }
    const mapping = {
        2: "abc",
        3: "def",
        4: "ghi",
        5: "jkl",
        6: "mno",
        7: "pqrs",
        8: "tuv",
        9: "wxyz",
    };
    let prefixSet = new Set([""]);
    for (const c of digits) {
        const newPrefixSet = new Set();
        const newDigits = mapping[c];
        for (const p of prefixSet) {
            for (const d of newDigits) {
                newPrefixSet.add(p + d);
            }
        }
        prefixSet = newPrefixSet;
    }
    return Array.from(prefixSet);
};
